DensityMatrix backend fixes

mgarnier59 opened the pull request:
Before submitting, please check the following:
- Make sure you have tests for the new code and that test passes (run `pytest`)
Passing

- format added code and tests by `black -l 120 <filename>`
Done

- If applicable, add a line to the [unreleased] part of CHANGELOG.md, following [keep-a-changelog](https://keepachangelog.com/en/1.0.0/).
 
Then, please fill in below:

**Context (if applicable):**
Density matrix backend.

**Description of the change:**

- Modified the normalization of density matrices from Frobenius (2-norm) to trace norm
- Added checks in the initialization of DensityMatrix to check for unit trace and hermiticity of the input data
- Modified the tests accordingly (normalization, hermiticity). For evolution tests, used unitary operators from `scipy.stats.unitary_group`. Can be modified in the future.

**Related issue:**
None.

also see that checks (github actions) pass.
If lint check keeps failing, try installing black==22.8.0 as behavior seems to vary across versions.

shinich1 commented on graphix/sim/density_matrix.py:
```diff
@@ -42,7 +43,14 @@ def __init__(self, data=None, plus_state=True, nqubit=1):
+            np.testing.assert_almost_equal(self.rho.trace(), 1, err_msg="The provided matrix does not have unit trace!")
+            # also add test for hermiticity!!! QuTip test? and modify the DM tests that use not hermitian DMs
+            np.testing.assert_equal(
+                scipy.linalg.ishermitian(self.rho), True, err_msg="The provided matrix is not Hermitian!"
```
how about using numpy, for example,
`np.testing.assert_allclose(self.rho, self.rho.getH())`?
scipy is not yet on requirements.txt.

shinich1 replied:
I see scipy is uesd for testing too; where necessary, we can install scipy specifically for testing, by adding to https://github.com/TeamGraphix/graphix/blob/master/.github/workflows/ci.yml

I've tried so far to keep requirements.txt as minimum as possible, for faster installation and less chance of version/compatibility issues in the future.

mgarnier59 replied:
> how about using numpy, for example, `np.testing.assert_allclose(self.rho, self.rho.getH())`? scipy is not yet on requirements.txt.

I didn't bother thinking about it first since I believed `scipy`was more efficient but I guess 'numpy' is a good way to go. 
Something that might be interesting for performance upgrade: Qutip mentions that it is possible to do a hermiticity check without compuing the conjugate transpose https://qutip.org/features.html. Might be worth having a look.



> I see scipy is uesd for testing too; where necessary, we can install scipy specifically for testing, by adding to https://github.com/TeamGraphix/graphix/blob/master/.github/workflows/ci.yml
> 
> I've tried so far to keep requirements.txt as minimum as possible, for faster installation and less chance of version/compatibility issues in the future.

Of course. It is just used to generate unitaries but it can be easily replace. Indeed from any random matrix building $U U^\dagger$ will yield a random unitary.

shinich1 replied:
The random unitary gen is for testing right? Perhaps the best way might be to require scipy for testing (as above), by adding to ci.yml - what do you think?

Of course, when it becomes necessary, I am always up to the idea of adding to requirements.txt too :) just trying to resist at first to keep the dependency as simple as possible.

mgarnier59 replied:
> The random unitary gen is for testing right? Perhaps the best way might be to require scipy for testing (as above), by adding to ci.yml - what do you think?

Yes, only for testing! I've added it to ci.yml, I hope I did it correctly. But does `pip install -e .[test]` install the additional dependencies specific to the tests?

shinich1 replied:
I was imagining adding `pip install scipy` to ci.yml but you had a good point here, it's actually here where we should add scipy:
https://github.com/TeamGraphix/graphix/blob/25e59bc5dfa0b3042348233d7b02626c84d2bf92/setup.py#L39C1-L39C77

mgarnier59 replied:
You sure it's better in the setup.py and not the ci? I guess tests are ran at install. 
Then what does `pip install -e .[test]`do if not installing the dependencies in the tests?

mgarnier59 commented:
New commit solving the issues we mentionned as well as a few updates:

- refactored 2-qubit gates application with `evolve` method (actually works for arbitrary number of qubits, tested for 1-, 2- and 3- qubits)
- rewrote the tensor contraction in a clearer way to avoid confusion (as I experienced) but kept the previous implementation as comment if need be.
- removed `scipy` from `density_matrix.py`
- added data validation with Exceptions instead of `assert`s
- added `scipy` and `random` to `setup.py`
- added tests according to new features

For the feature, maybe generic data validation like checking operators are square matrices of size $2^n \times 2^N$ can be mutualised with the density matrix that just adds further conditions on top of that (hermiticity, unit trace). Maybe put those methods in a specific file?

shinich1 commented:
Thanks! Will take a look through later. 

> Maybe put those methods in a specific file?
good idea, sounds like a portable feature outside this simulator. maybe a new file in in `graphix/sim/`?

mgarnier59 commented:
> good idea, sounds like a portable feature outside this simulator. maybe a new file in in `graphix/sim/`?

Actually I was thinking of it while working on the `kraus.py`file that requires the same types of checks. So either we move `kraus` (we might rename it) in `graphix/sim/` or we include the checks in `graphix/` but that doesn't really matter much.

shinich1 commented:
@mgarnier59 maybe graphix/sim/quantum_channel.py, but as you say it does not matter much at this stage. 
let's continue with whatever that works for you, we then deal with file arrangement after all the implementation part is ready :)

shinich1 commented on graphix/sim/density_matrix.py:
```diff
@@ -66,6 +72,73 @@ def evolve_single(self, op, i):
+        )
+        self.rho = rho_tensor.reshape((2**self.Nqubit, 2**self.Nqubit))
+
+        # # Original and alternative way : tensor transpose instead of matrix (QM) transpose
+
+        # op_tensor = op.reshape((2,) * 2 * op_dim)
+
+        # rho_tensor = self.rho.reshape((2,) * self.Nqubit * 2)
+
+        # rho_tensor = np.tensordot(
+        #     np.tensordot(op_tensor, rho_tensor, axes=[tuple(op_dim + i for i in range(len(qargs))), tuple(qargs)]),
+        #     op_tensor.conj().T,
+        #     axes=[tuple(i + self.Nqubit for i in reversed(qargs)), tuple(i for i in range(len(qargs)))],
+        # )
+        # rho_tensor = np.moveaxis(
+        #     rho_tensor,
+        #     [i for i in range(len(list(qargs)))] + [-i for i in range(1, len(list(qargs)) + 1)],
+        #     [i for i in qargs] + [i + self.Nqubit for i in list(qargs)],
+        # )
+        # self.rho = rho_tensor.reshape((2**self.Nqubit, 2**self.Nqubit))
```
let's get rid of this!

mgarnier59 replied:
Done.

shinich1 commented on graphix/sim/density_matrix.py:
```diff
@@ -91,20 +164,24 @@ def cnot(self, edge):
-        )
-        rho_tensor = np.moveaxis(rho_tensor, (0, 1, -1, -2), (edge[0], edge[1], edge[0] + n, edge[1] + n))
-        self.rho = rho_tensor.reshape((2**n, 2**n))
+        # already an attribute
+        # n = int(np.log2(self.rho.shape[0]))
+        # assert i >= 0 and j >= 0
+        # assert self.Nqubit > i and self.Nqubit > j
+        # assert i != j
+
+        # rho_tensor = self.rho.reshape((2,) * n * 2)
+        # rho_tensor = np.tensordot(
+        #     np.tensordot(CNOT_TENSOR, rho_tensor, axes=[(2, 3), edge]),
+        #     CNOT_TENSOR.conj().T,
+        #     axes=[(edge[1] + n, edge[0] + n), [0, 1]],
+        # )
+        # rho_tensor = np.moveaxis(rho_tensor, (0, 1, -1, -2), (edge[0], edge[1], edge[0] + n, edge[1] + n))
+        # self.rho = rho_tensor.reshape((2**n, 2**n))
```
this too

mgarnier59 replied:
Done.

shinich1 commented on graphix/sim/density_matrix.py:
```diff
@@ -218,6 +272,7 @@ def __init__(self, pattern, max_qubit_num=12):
         if pattern.max_space() > max_qubit_num:
             raise ValueError("Pattern.max_space is larger than max_qubit_num. Increase max_qubit_num and try again.")
 
+    # TODO to be removed
     def dephase(self, p=0):
         """Apply dephasing channel to all nodes. Phase is flipped with probability p.
 
```
let's remove now?

mgarnier59 replied:
If that's ok with you, I plan to remove it in the next contribution that incorporates the noise handling

shinich1 commented on tests/test_density_matrix.py:
```diff
@@ -455,6 +628,44 @@ def qft_circ():
+
+
 if __name__ == "__main__":
     np.random.seed(2)
```
perhaps better to remove seeding?

mgarnier59 replied:
I think it was there before I jumped in. Maybe it's better for reproducibility?

shinich1 commented on .gitignore:
```diff
@@ -6,3 +6,4 @@
 **/build
 dist/*
 docs/source/gallery/**
+.pre-commit-config.yaml
```
let's get rid of this?

mgarnier59 replied:
Done.

shinich1 reviewed (approved):
There are a few parts that needs tidying up, but looks good otherwise!
Let's get rid of pre-commit-config from gitignore too. 
After that, please squash and merge.

shinich1 commented:
@mgarnier59 I just assigned you to the PR (sorry I haven't done this yet). Do you see the button for squash and merge now?
